server/server.go

```diff
@@ -42,72 +55,362 @@ func HandleWebsocketConnection(serverHTTPResponse http.ResponseWriter, clientHTT
 	if websocketConnectionErr != nil {
 		// If there was an error while upgrading the HTTP connection to a WebSocket connection return a relevant error
 		http.Error(serverHTTPResponse, "Unable to upgrade HTTP connection to a WebSocket connection", 400)
+		// Close the websocket connection
+		websocketConnection.Close()
 		return
 	}
-
 	// Require successful authentication before allowing a client to send a message
-	_, websocketConnectionStateErr := requireAuth(websocketConnection)
-	if websocketConnectionStateErr != nil {
+	authenticatedIdentityPublicKeyHexClient, websocketConnectionrequestAuthErr := requestAuth(websocketConnection)
+	if websocketConnectionrequestAuthErr != nil {
 		// If the authentication failed, terminate the websocket connection to the client
-		log.Println("Authentication Failed:", websocketConnectionStateErr)
-		websocketConnection.WriteMessage(websocket.BinaryMessage, []byte(websocketConnectionStateErr.Error()))
+		log.Println("Authentication Failed:", websocketConnectionrequestAuthErr)
+		websocketConnection.WriteMessage(websocket.BinaryMessage, []byte(websocketConnectionrequestAuthErr.Error()))
 		log.Println("Terminating websocket connection to client.")
+		// Close the websocket connection
+		websocketConnection.Close()
 		return
 	}
+	// Check if there are pending messages to be delivered when the client comes back online
+	if messagesToBeDelivered, ok := multiUserChatMessage[authenticatedIdentityPublicKeyHexClient]; ok {
+		deliverMessages(websocketConnection, messagesToBeDelivered)
+	} // if messagesToBeDelivered
 
-	// Notify the client that authentication succeeded
-	websocketConnection.WriteMessage(websocket.BinaryMessage, []byte("Authentication Succeeded"))
-
+	websocketConnectionProcessMessageErr := processMessage(websocketConnection, authenticatedIdentityPublicKeyHexClient)
+	for websocketConnectionProcessMessageErr == nil {
+		websocketConnectionProcessMessageErr = processMessage(websocketConnection, authenticatedIdentityPublicKeyHexClient)
+	}
+	// Close the websocket connection
+	websocketConnection.Close()
+	return
+	// Read first message from client
 } // func handleWebsocketConnection
 
-func requireAuth(websocketConnection *websocket.Conn) (int, error) {
-	// Create a variable to keep track of the websocket connection state
-	// Create a byte sequence that we would ask the client to sign
-	byteSequenceToSign := []byte("hi")
-	// Initialize the Auth protobuf structure
-	protobufAuth := backendProtobuf.Auth{}
-	// Fill in the byte sequence that we would ask the client to sign
-	protobufAuth.ToSign = byteSequenceToSign
+func processMessage(websocketConnection *websocket.Conn, authenticatedIdentityPublicKeyHexClient string) error {
```

Comment:
`authenticatedIdentityPublicKeyHexClient` is currently not used.

Reply:
Noted.